### PR TITLE
feat(typography): introduce CodeBlock component

### PR DESCRIPTION
**Pull request by jzempel**
## Description

The new `CodeBlock` component statically renders blocks of code with language syntax highlighting based on [Prism](https://prismjs.com/) and with Garden color palette theming.

## Detail

After careful evaluation, I selected [prism-react-renderer](https://github.com/FormidableLabs/prism-react-renderer) due to flexibility of the model, structural semantics (block -> line -> token), and reduced interference with the underlying Prism tokenization. While more popular, [highlight.js](https://highlightjs.org/) proved to be problematic on each of these points.

There are still plenty of styling nits with the final result, but we obtained the best result that the underlying technology is able to give us. Ideally, Prism improves over time, and our `CodeBlock` will improve with it.

## Checklist

- [x] :ok_hand: design updates are Garden Designer approved (add the
      designer as a reviewer)
- [x] :globe_with_meridians: Styleguidist demo is up-to-date (`yarn start`)
- [x] :arrow_left: renders as expected with reversed (RTL) direction
- [x] :metal: renders as expected with Bedrock CSS (`?bedrock`)
- [x] :wheelchair: analyzed via [axe](https://www.deque.com/axe/) and evaluated using VoiceOver
- [x] :guardsman: includes new unit tests
- [x] :memo: tested in Chrome, Firefox, Safari, Edge, and IE11

### CODE REVIEW

File: packages/typography/src/styled/StyledCodeBlock.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+ * found at http://www.apache.org/licenses/LICENSE-2.0.
+ */
+
+import styled, { ThemeProps, DefaultTheme, css } from 'styled-components';
+import { DEFAULT_THEME, getColor, retrieveComponentStyles } from '@zendeskgarden/react-theming';
+
+const COMPONENT_ID = 'typography.codeblock';
+
+const colorStyles = (props: IStyledCodeBlockProps & ThemeProps<DefaultTheme>) => {
+  const backgroundColor = getColor('neutralHue', props.isLight ? 100 : 1000, props.theme);
```
Comment (jzempel) [inline packages/typography/src/styled/StyledCodeBlock.ts:14]: @steue note that the use of `1000` here awaits the decision on `grey-900`.

File: packages/typography/src/elements/CodeBlock.tsx [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+  isLight?: boolean;
+  isNumbered?: boolean;
+}
+
+/**
+ * @extends HTMLAttributes<HTMLPreElement>
+ */
+const CodeBlock: React.FunctionComponent<
+  ICodeBlockProps & React.RefAttributes<HTMLPreElement>
+> = React.forwardRef(({ children, isLight, isNumbered, language, size, ...other }, ref) => {
```
Comment (austin94) [inline packages/typography/src/elements/CodeBlock.tsx:24]: `React.forwardRef<T, P>()` takes two generics which can remove the need for `React.RefAttributes` and `React.FunctionComponent`.

**Review by austin94 — CHANGES_REQUESTED** [review]

File: packages/typography/src/styled/StyledCodeBlock.ts (hunk 1 above)
Comment (steue) [reply packages/typography/src/styled/StyledCodeBlock.ts:14]: @jzempel design team's consensus is to use `getColor`, as `grey-900` one off isn't ideal and should be visited holistically.

Comment (jzempel) [reply packages/typography/src/styled/StyledCodeBlock.ts:14]: Awesome. I'm a bit relieved.

File: packages/typography/src/elements/CodeBlock.tsx (hunk 2 above)
Comment (jzempel) [reply packages/typography/src/elements/CodeBlock.tsx:24]: At present, it seems we have dozens of declarations scattered throughout the codebase that don't line up with the simplified syntax. Would probably be good to set a task to clean those up so folks (like me 😅 ) don't keep following suboptimal examples.

**Review by austin94 — APPROVED** [review]

**Review by hzhu — APPROVED** [review]

**Review by steue — APPROVED** [review]
so good! thank you @jzempel !!!!

**Pull request merged**